### PR TITLE
Create users service

### PR DESCRIPTION
**Pull request by alealvarez14**
## Pull request type

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

- Auth controller has a lot of logic as well as users entity. 

Issue Link: [Create users service](https://github.com/rootstrap/node-ts-api-base/projects/1#card-52906465)

## What is the new behavior?

- Create the users service to wrap all the logic like comparing passwards, creation of token, hashing password, etc

## Other information

N/A

## Preview

N/A

### CODE REVIEW

File: src/services/users.service.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { Service } from 'typedi';
+import { compareSync, genSaltSync, hashSync } from 'bcrypt';
+import { User } from '@entities/user.entity';
+import { JWTService } from '@services/jwt.service';
+
+@Service()
+export class UsersService {
+  constructor(private jwtService: JWTService) {}
+
+  credentialsExists(email: string, password: string): boolean {
```
Comment (Alxzu) [inline src/services/users.service.ts]:
```suggestion
  credentialsExists(email: string, password: string): boolean {
      return email && password
```

File: src/controllers/auth.controller.ts [hunk 2]
```diff
@@ -33,7 +33,7 @@ export class AuthController {
     @BodyParam('email') email: string,
     @BodyParam('password') password: string
   ) {
-    if (!email || !password) {
+    if (!this.usersService.credentialsExists(email, password)) {
```
Comment (Alxzu) [inline src/controllers/auth.controller.ts]: What about renaming this like `givenCredentials`? Exists makes me thing we are checking that they appear on the DB, wdyth?

**Review by martinjaimem — APPROVED** [review]

File: src/services/users.service.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    return !!(email && password);
+  }
+
+  comparePassword(password: string, userPass: string): boolean {
+    return compareSync(password, userPass);
+  }
+
+  async generateToken(user: User) {
+    return await this.jwtService.createJWT(user);
+  }
```
Comment (martinjaimem) [inline src/services/users.service.ts]:
```suggestion
  generateToken(user: User) {
    return this.jwtService.createJWT(user);
  }
```

**Review by martinjaimem — APPROVED** [review]

**Review by kevquincke — APPROVED** [review]

**Review by Alxzu — APPROVED** [review]

**Review by kevquincke — APPROVED** [review]
Good one! 👏

File: src/controllers/auth.controller.ts [hunk 4]
```diff
@@ -7,52 +7,42 @@ import {
...
   async signUp(@Body({ validate: false }) user: User, @Res() response: any) {
-    let newUser;
-
     try {
-      newUser = await getRepository(User).save(user);
+      const newUser = await this.usersService.signUp(user);
+      return response.send(_.omit(newUser, ['password']));
     } catch (error) {
-      throw new BadRequestError('Missing params on body');
+      if (error?.message === Errors.MISSING_PARAMS) {
```
Comment (martinjaimem) [inline src/controllers/auth.controller.ts:25]: 👏

File: src/services/users.service.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
+import { Service } from 'typedi';
+import { compareSync, genSaltSync, hashSync } from 'bcrypt';
+import { getRepository } from 'typeorm';
+import { User } from '@entities/user.entity';
+import { JWTService } from '@services/jwt.service';
+import { Errors } from '@constants/errorMessages';
+
+@Service()
```
Comment (martinjaimem) [inline src/services/users.service.ts:7]: Nice 😎 , I'd (personally) make some changes to the service, I'm not a huge fan of services like `UserService` as IMHO a service should have one specific goal, for example, the sign-in of a user (SessionService).

Moreover, the methods such as `index, post, put, delete show` I think that could be removed from the service as they don't really have much logic for themselves, they are one-line methods that do the job needed so they (again, IMO) could be in the controller.

File: src/services/users.service.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    try {
+      user = await User.findOneOrFail({ where: { email } });
+    } catch (error) {
+      throw new Error('UnauthorizedError');
+    }
+
+    // Check if encrypted password match
+    if (!this.comparePassword(password, user.password)) {
+      throw new Error('Invalid credentials');
+    }
```
Comment (martinjaimem) [inline src/services/users.service.ts]: I think the method is really well written, no need for comments in the code

File: src/services/users.service.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    // user matches email + password, create a token
```
Comment (martinjaimem) [inline src/services/users.service.ts]: same here

File: src/services/users.service.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      throw new Error(Errors.MISSING_PARAMS);
+    }
+
+    return newUser;
+  }
+
+  async signIn(email: string, password: string) {
+    if (!this.givenCrentials(email, password)) {
+      throw new Error('Missing params on body');
+    }
```
Comment (martinjaimem) [inline src/services/users.service.ts]: Seems to be logic associated with the request and not too much with the login. The service doesn't need to know that a request was done, it should just check if the credentials are correct, and raise an error if they are not, then the controller can decide how to render that error, but shouldn't be the service responsibility

**Review by martinjaimem — COMMENTED** [review]
Niceee,

Would be awesome if we added some tests too 😎

**Review by pablanco — APPROVED** [review]
Nice!

**Pull request merged**